[Mengzhe] iP

charliemoweng opened the pull request:
# DukeAzure
> “Focus on being productive instead of busy.” - Tim Ferriss, American podcaster, author and entrepreneur

Here at DukeAzure, we help you be productive so you don't have to be busy taking down your tasks everywhere, ever.

### Some selling points include:
- easy to use
- small-sized
- ~~fast~~ *VERY* fast
- Permanently **FREE**

Get yours from [here](https://github.com/charliemoweng/ip/releases/tag/v0.1.1) today!

### Some simple steps to follow:
1. Follow the link above to the webpage.
2. Download it!
3. Run it in your Com :computer: !
4. Start adding your tasks and manage them as you wish.

### Some features (many more are on the way)
- [X] Add and store a task
- [X] Mark a task as done
- [X] Show all your tasks
- [X] Find a specific task
- [ ] Send reminders before a task is due

### For those avid coders out there
Here is a snippet of the code. Feel free to give suggestions!
The `main` method of Duke:

```java
public static void main(String[] args) {
       new Duke(FILE_PATH).run();
} 
```

Commit:
Add Gradle support

charliemoweng committed:
test hi

charliemoweng committed:
Level-1

charliemoweng committed:
Level-2: Add, List

charliemoweng committed:
Level 3: mark as done. Added in Task class.

charliemoweng committed:
Duke: updated method used in 'set task as Done' to using regex.
Task: added toString() method.
Added Todo, Deadline and Event classes.

charliemoweng committed:
Level 4. ToDos, Events, Deadlines

charliemoweng committed:
testing

charliemoweng committed:
Level 5: minimal, yet to implement custom Exception class

charliemoweng committed:
Level 6: added the function to delete a task

charliemoweng committed:
test push

charliemoweng committed:
test push done

charliemoweng committed:
Add the function to save data to local

charliemoweng committed:
Change the date time of Deadline and Event from string to LocalDateTime

charliemoweng committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Change the date time of Deadline and Event from string to LocalDateTime

# Conflicts:
#	src/main/java/Duke.java

Resolve the conflict

charliemoweng committed:
Add in the function to read from data.txt

charliemoweng committed:
Remove commented out helper function (no longer used)

jackgugz commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,20 @@
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.time.temporal.ChronoUnit;
+
```
We need to add some JavaDoc comments for all public classes and methods. Noticed the lack of javadoc in some other places too.

charliemoweng replied:
Yes indeed. I will add that in soon. Thank you for pointing it out.

jackgugz commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,350 @@
+import java.io.IOException;
+import java.io.FileNotFoundException;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
```
Should an empty line be added between imports from different packages? Also shall we order imports from same package by their alphabetical order?

charliemoweng replied:
Yes, you are right. I will reorder them to make them look nicer.

jackgugz commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,350 @@
 public class Duke {
+
+    private static void appendToFile(String filePath, String textToAppend) throws IOException {
+        FileWriter fw = new FileWriter(filePath, true); // create a FileWriter in append mode
```
Should comments be indented relative to their position in the code?
```suggestion
        // create a FileWriter in append mode
        FileWriter fw = new FileWriter(filePath, true); 
```

charliemoweng replied:
you are right.

jackgugz commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,350 @@
+//            System.out.println(s.nextLine());
+//        }
+//    }
+
```
Shall we avoid such commented out codes?

jackgugz commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,350 @@
+                    System.out.println("Data file has been created.");
+                } else {
+                    // System.out.println("Data folder is already present.");
+                }
```
Is it possible that we take out the else clause then?

charliemoweng replied:
yes thank you for pointing it out. i have removed it since.

jackgugz commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,350 @@
+                File dataDirectory = new File("./data/data.txt");
+                if (dataDirectory.createNewFile()) {
+                    System.out.println("Data file has been created.");
+                } else {
+                    // System.out.println("Data folder is already present.");
+                }
```
Same as above

jackgugz commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,350 @@
+            System.out.println("data.txt is not found.");
+        }
+
+//        for (int i = 0; i < inputs.size(); i++) {
+//            System.out.println("test read todo and ddl\n" + dataRead.get(i));
+//        }
```
We may delete this as well.

jackgugz commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,350 @@
+            // mark a task as Done
+            Pattern pattern = Pattern.compile("done\\s\\d+");
+            Matcher matcher = pattern.matcher(input);
+            boolean matchFound = matcher.find();
```
```suggestion
            boolean ifMatchFound = matcher.find();
```
so that it sounds more like a boolean

jackgugz commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,350 @@
+            // delete a task
+            Pattern patternDelete = Pattern.compile("delete\\s\\d+");
+            Matcher matcherDelete = patternDelete.matcher(input);
+            boolean matchFoundDelete = matcherDelete.find();
```
Same to the boolean name here as well.

jackgugz commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,350 @@
+                // Case Event
+                if (group[0].equals(taskEve) && input.contains(" /at ")) {
+                    String[] eveGroup = input.split(" /at ");
+                    String eveToAdd = eveGroup[0].substring(6); // name of the task is after "event" and space
```
```suggestion
                    // name of the task is after "event" and space
                    String eveToAdd = eveGroup[0].substring(6); 
```

jackgugz commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,19 @@
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.time.temporal.ChronoUnit;
```
The ordering of the imports can be better!

charliemoweng replied:
yes i will improve upon it.

jackgugz commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,23 @@
+     }
+
+    public String getStatusIcon() {
+        return (isDone ? "X" : " "); // mark done task with X
```
```suggestion
        // mark done task with X
        return (isDone ? "X" : " "); 
```

jackgugz reviewed (commented):
Overall, I see consistent good coding standards being especially for the `for` loops and conditional statements! The spacing and indentation made it super easy to read. However, I do observe some common places for improvements: 
1. ordering of the imports
2. comments indentation and lack of javadoc comments
3. naming of variables. 

Nonetheless still good job! Looking forward to the more updated version as I think it may be a few versions behind the schedule.

1waykiat commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,350 @@
+            }
+        }
+
+        // if data.txt does not exist, create file
+        if (!dataFile.exists()) {
+            try {
+                File dataDirectory = new File("./data/data.txt");
+                if (dataDirectory.createNewFile()) {
+                    System.out.println("Data file has been created.");
+                } else {
+                    // System.out.println("Data folder is already present.");
+                }
+            } catch (IOException e) {
+                System.out.println("An error occurred.");
+                e.printStackTrace();
+            }
+        }
```
Could be extracted out to the Storage class to handle instead

1waykiat commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,350 @@
+            Pattern pattern = Pattern.compile("done\\s\\d+");
+            Matcher matcher = pattern.matcher(input);
+            boolean matchFound = matcher.find();
+            if (matchFound) {
```
Perhaps let Parser class handle the match finding instead, and extract the code there.

1waykiat reviewed (commented):
Remember to complete the tasks from last week! 
Mostly just need to extract away the Storage and Parser class so that the parsing and loading and storing of data can be abstracted away.

charliemoweng committed:
Implement more OOP: create classes Ui, Storage, Parser and TaskList.

charliemoweng committed:
Organise classes into packages

charliemoweng committed:
Change file path of data.txt

charliemoweng committed:
Add JUnit tests for Parser.java, Task.java and Todo.java

charliemoweng committed:
Move Duke.java out of duke package, change import statement from wildcard to specific, create JAR file

charliemoweng committed:
put Duke.java into package Duke, change path of data.txt

charliemoweng committed:
write javadoc for all classes

charliemoweng committed:
check code style

charliemoweng committed:
add find method

charliemoweng committed:
add find method and others

charliemoweng committed:
Merge commit 'd8398594b7bc43da5eb865321c5a50cec4b3923d' into branch-A-Gradle

* commit 'd8398594b7bc43da5eb865321c5a50cec4b3923d':
  Add Gradle support

charliemoweng committed:
add in checkstyle rules and check style

charliemoweng committed:
change from CLI to GUI

charliemoweng committed:
use Gradle to run the application

charliemoweng committed:
Add assert feature

Assertions are a crucial part of Java error handling since.
It helps detects errors at runtime and terminates the application.

Assertions are missing in this project.

Add assert features in TaskList.java because it has many if blocks
that are suitable for using assert.

charliemoweng committed:
Improve code quality

Good code quality denotes that methods are not too long, not beyond
one screen.

Some methods in Storage.java and TaskList.java are too long.

It is hard to read and locate specific functionalities.

Extract the logic and create helper methods within its class.

charliemoweng committed:
Merge pull request #2 from charliemoweng/branch-A-Assertions

Add assert feature

charliemoweng committed:
Merge branch 'master' into branch-A-CodeQuality

* master:
  Add assert feature

charliemoweng committed:
Merge pull request #3 from charliemoweng/branch-A-CodeQuality

Improve code quality

charliemoweng committed:
Add the functionality to check for duplicate tasks

We assume that users would not want to input the same task twice.
This prevents possible confusion.

TaskList allows the creation of duplicates.

Add in methods to check for duplicating Todo, Deadline, and Event.
Only add one if it is not a duplicate.

This enhances user experience.

charliemoweng committed:
Adjust GUI features

charliemoweng committed:
Add acknowledgement for code reused

Find code to make Label of DialogBox have infinite height to
accommodate longer responses on GitHub and CS2103 Forum.

Acknowledge it.

charliemoweng committed:
Change comment to correct format

charliemoweng committed:
Add in representative screenshot

charliemoweng committed:
Modify switch statement to comply with style

charliemoweng committed:
Remove JavaFX FXML API version number

It shows warning of incompatible version numbers when the application
is executed.

Remove the version number from fxml files.

charliemoweng committed:
Add in a check for data file

Data file under the default path of ./data/data.txt
may or may not exist.
There is no need to create it if it already exists.

The Storage class always creates the data file.
If the file already exists, it will enter the catch block and
print out error message when there is no error at all.

Add in a check before creating the file. Only create it when
it is not yet existent.

charliemoweng committed:
Update README.md

charliemoweng committed:
Update README.md

Should be the finalised version of README

charliemoweng committed:
Add in welcome message on start up

charliemoweng committed:
Format a few responses to change to a new line

charliemoweng committed:
Remove javafx version number

charliemoweng committed:
Declare constant Strings and switch order to pass style check

charliemoweng committed:
Update README.md

charliemoweng committed:
Set theme jekyll-theme-minimal

charliemoweng committed:
Update README.md
